File: app/src/test/java/org/thoughtcrime/securesms/database/CursorRecyclerViewAdapterTest.java
package org.thoughtcrime.securesms.database;

import android.content.Context;
import android.database.Cursor;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;
import android.view.View;
import android.view.ViewGroup;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.*;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class CursorRecyclerViewAdapterTest {
  private CursorRecyclerViewAdapter adapter;
  private Context                   context;
  private Cursor                    cursor;

  @Before
  public void setUp() {
    context = mock(Context.class);
    cursor  = mock(Cursor.class);
    when(cursor.getCount()).thenReturn(100);
    when(cursor.moveToPosition(anyInt())).thenReturn(true);

    adapter = new CursorRecyclerViewAdapter(context, cursor) {
      @Override
      public ViewHolder onCreateItemViewHolder(ViewGroup parent, int viewType) {
        return null;
      }

      @Override
      public void onBindItemViewHolder(ViewHolder viewHolder, @NonNull Cursor cursor) {
      }
    };
  }

  @Test
  public void testSanityCount() throws Exception {
    assertEquals(adapter.getItemCount(), 100);
  }

  @Test
  public void testHeaderCount() throws Exception {
    adapter.setHeaderView(new View(context));
    assertEquals(adapter.getItemCount(), 101);

    assertEquals(adapter.getItemViewType(0), CursorRecyclerViewAdapter.HEADER_TYPE);
    assertNotEquals(adapter.getItemViewType(1), CursorRecyclerViewAdapter.HEADER_TYPE);
    assertNotEquals(adapter.getItemViewType(100), CursorRecyclerViewAdapter.HEADER_TYPE);
  }

  @Test
  public void testFooterCount() throws Exception {
    adapter.setFooterView(new View(context));
    assertEquals(adapter.getItemCount(), 101);
    assertEquals(adapter.getItemViewType(100), CursorRecyclerViewAdapter.FOOTER_TYPE);
    assertNotEquals(adapter.getItemViewType(0), CursorRecyclerViewAdapter.FOOTER_TYPE);
    assertNotEquals(adapter.getItemViewType(99), CursorRecyclerViewAdapter.FOOTER_TYPE);
  }

  @Test
  public void testHeaderFooterCount() throws Exception {
    adapter.setHeaderView(new View(context));
    adapter.setFooterView(new View(context));
    assertEquals(adapter.getItemCount(), 102);
    assertEquals(adapter.getItemViewType(101), CursorRecyclerViewAdapter.FOOTER_TYPE);
    assertEquals(adapter.getItemViewType(0), CursorRecyclerViewAdapter.HEADER_TYPE);
    assertNotEquals(adapter.getItemViewType(1), CursorRecyclerViewAdapter.HEADER_TYPE);
    assertNotEquals(adapter.getItemViewType(100), CursorRecyclerViewAdapter.FOOTER_TYPE);
  }
}

File: app/src/test/java/org/thoughtcrime/securesms/database/CursorRecyclerViewAdapterTest.kt
package org.thoughtcrime.securesms.database

import android.content.Context
import android.database.Cursor
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import io.mockk.every
import io.mockk.mockk
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Test
import org.thoughtcrime.securesms.database.CursorRecyclerViewAdapter.FOOTER_TYPE
import org.thoughtcrime.securesms.database.CursorRecyclerViewAdapter.HEADER_TYPE

class CursorRecyclerViewAdapterTest {
  private val context: Context = mockk<Context>()
  private val cursor: Cursor = mockk<Cursor>(relaxUnitFun = true) {
    every { count } returns 100
    every { moveToPosition(any()) } returns true
  }
  private val adapter = object : CursorRecyclerViewAdapter<RecyclerView.ViewHolder?>(context, cursor) {
    override fun onCreateItemViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder? = null
    override fun onBindItemViewHolder(viewHolder: RecyclerView.ViewHolder?, cursor: Cursor) = Unit
  }

  @Test
  fun testSanityCount() {
    assertEquals(100, adapter.itemCount)
  }

  @Test
  fun testHeaderCount() {
    adapter.headerView = View(context)
    assertEquals(101, adapter.itemCount)

    assertEquals(HEADER_TYPE, adapter.getItemViewType(0))
    assertNotEquals(HEADER_TYPE, adapter.getItemViewType(1))
    assertNotEquals(HEADER_TYPE, adapter.getItemViewType(100))
  }

  @Test
  fun testFooterCount() {
    adapter.setFooterView(View(context))
    assertEquals(101, adapter.itemCount)
    assertEquals(FOOTER_TYPE, adapter.getItemViewType(100))
    assertNotEquals(FOOTER_TYPE, adapter.getItemViewType(0))
    assertNotEquals(FOOTER_TYPE, adapter.getItemViewType(99))
  }

  @Test
  fun testHeaderFooterCount() {
    adapter.headerView = View(context)
    adapter.setFooterView(View(context))
    assertEquals(102, adapter.itemCount)
    assertEquals(FOOTER_TYPE, adapter.getItemViewType(101))
    assertEquals(HEADER_TYPE, adapter.getItemViewType(0))
    assertNotEquals(HEADER_TYPE, adapter.getItemViewType(1))
    assertNotEquals(FOOTER_TYPE, adapter.getItemViewType(100))
  }
}
